pkg/framework/objects/service_token/resource.go

```diff
@@ -0,0 +1,325 @@
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/int64default"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/setdefault"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+var (
+	_ resource.Resource                = &serviceTokenResource{}
+	_ resource.ResourceWithConfigure   = &serviceTokenResource{}
+	_ resource.ResourceWithImportState = &serviceTokenResource{}
+)
+
+func ServiceTokenResource() resource.Resource {
+	return &serviceTokenResource{}
+}
+
+type serviceTokenResource struct {
+	client *dbt_cloud.Client
+}
+
+// Metadata implements resource.Resource.
+func (st *serviceTokenResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_service_token"
+}
+
+// Configure implements resource.ResourceWithConfigure.
+func (st *serviceTokenResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	switch c := req.ProviderData.(type) {
+	case nil: // do nothing
+	case *dbt_cloud.Client:
+		st.client = c
+	default:
+		resp.Diagnostics.AddError("Missing client", "A client is required to configure the service token resource")
+	}
+}
+
+// Schema implements resource.Resource.
+func (st *serviceTokenResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed:    true,
+				Description: "The ID of the service token",
+				// this is used so that we don't show that ID is going to change
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"uid": schema.StringAttribute{
+				Description: "Service token UID (part of the token)",
+				Computed:    true,
+			},
+			"name": schema.StringAttribute{
+				Description: "Service token name",
+				Required:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"token_string": schema.StringAttribute{
+				Description: "Service token secret value (only accessible on creation))",
+				Computed:    true,
+				Sensitive:   true,
+			},
+			"state": schema.Int64Attribute{
+				Description: "Service token state (1 is active, 2 is inactive)",
+				Optional:    true,
+				Computed:    true,
+				Default:     int64default.StaticInt64(1),
+			},
+		},
+		Blocks: map[string]schema.Block{
+			"service_token_permissions": schema.SetNestedBlock{
```

Comment:
That's great that you kept a `SetNestedBlock` for backward compatibility. 👍 

For new resources or new nested attributes for existing resources Hashicorp recommends to move from [Nested Blocks](https://developer.hashicorp.com/terraform/plugin/framework/handling-data/blocks/set-nested) to [Nested Attributes ](https://developer.hashicorp.com/terraform/plugin/framework/handling-data/attributes#nested-attribute-types) but I don't think that we should move the SDK resources to the new type as the configuration format changes and we would break existing config/modules.

Reply:
Yes, I would have used a NestedAttribute here, but the SDKv2 version used a block. This was 100% for backwards compatibility. I aimed for 1:1 parity with the old resource.